instrumentation/apache-log4j-2/src/main/java/org/apache/logging/log4j/core/LogEvent_Instrumentation.java

```diff
@@ -0,0 +1,80 @@
+/*
+ *
+ *  * Copyright 2023 New Relic Corporation. All rights reserved.
+ *  * SPDX-License-Identifier: Apache-2.0
+ *
+ */
+
+package org.apache.logging.log4j.core;
+
+import com.newrelic.api.agent.NewRelic;
+import com.newrelic.api.agent.weaver.MatchType;
+import com.newrelic.api.agent.weaver.NewField;
+import com.newrelic.api.agent.weaver.Weave;
+import org.apache.logging.log4j.Level;
+import org.apache.logging.log4j.Marker;
+import org.apache.logging.log4j.ThreadContext;
+import org.apache.logging.log4j.core.impl.ThrowableProxy;
+import org.apache.logging.log4j.core.time.Instant;
+import org.apache.logging.log4j.message.Message;
+import org.apache.logging.log4j.util.ReadOnlyStringMap;
+
+import java.util.Map;
+
+import static com.newrelic.agent.bridge.logging.AppLoggingUtils.isApplicationLoggingLocalDecoratingEnabled;
+
+@Weave(originalName = "org.apache.logging.log4j.core.LogEvent", type = MatchType.Interface)
+public abstract class LogEvent_Instrumentation {
+
+    /*
+     * In cases where the LogEvent is sent to an AsyncAppender, getLinkingMetadata would get called on a new thread and the trace.id and span.id
+     * would be missing. To work around this we save the linking metadata on the LogEvent on the thread where it was created and use it later.
+     */
+    @NewField
+    public Map<String, String> agentLinkingMetadata = isApplicationLoggingLocalDecoratingEnabled() ? NewRelic.getAgent().getLinkingMetadata() : null;
+
+    public abstract LogEvent toImmutable();
```

Comment:
I didn't think they were required but oddly when I removed them, leaving the `@NewField` as the only thing in the interface, the instrumentation no longer worked. I expected that abstract methods in the interface didn't matter unless we were explicitly weaving them but the agent behavior seemed to indicate otherwise. I don't fully understand why. If that wasn't the case then this instrumentation wouldn't have to be split into two modules.